File: homework/Manh_Nguyen_Van/lesson-1/206-reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode preListNode= null;
        ListNode temp= null;
        return reverseByIterator(head, preListNode, temp);
    }
    
    public ListNode reverseByIterator(ListNode head, ListNode preListNode, ListNode temp) {
		while (head != null) {
			temp = head.next;
			head.next = preListNode;
			preListNode = head;
			if (temp == null) {
				break;
			}
			head = temp;
		}
		return head;
	}

	public ListNode reverseByRecursive(ListNode head, ListNode preListNode, ListNode temp) {
		if (head == null) {
			return head;
		}
		temp = head.next;
		head.next = preListNode;
		preListNode = head;
		if (temp == null) {
			return head;
		}
		head = temp;
		return reverseByRecursive(head, preListNode, temp);
	}
}

File: homework/Truong_Duong/lesson_1/206_Reverse_linked_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    # Iterative approach
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev, current, next = None, head, None

        while(current != None):
            next = current.next
            current.next = prev
            prev = current
            current = next

        return prev

    # Recursive approach
    def reverseListRecursive(self, head: Optional[ListNode]) -> Optional[ListNode]:
        def reverseListUtil(head, next):
            if head == None: return next
            temp = head.next
            head.next = next
            return reverseListUtil(temp, head)

        return reverseListUtil(head, None)